### PR TITLE
Compat: header validation refactor and tests

### PR DESCRIPTION
**Pull request by alexandre-merle**
Refactor and tests the header validation to ensure compatibility against real AWS S3 api

1) Modularizes the validateHeaders function so that 3 functions that were combined are separated: (a) checking whether the object exists, (b) checking the if-match and if-modified headers against the object metadata, (c) pulling the metadata for the response.  This impacts objectHead, deleteObject and getObject API's.

2) Fixes the if-match and if-modified header checks to deal with multiple etags provided in one request and for AWS compatibility with mixed headers (e.g., if-match and if-modified-since together in the same request).  This closes https://github.com/scality/S3/issues/125

### CODE REVIEW

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+            s3.headObject(Object.assign({
+                Bucket: bucketName,
+                Key: objectName,
+            }, fields), cb);
+        }
+
+        beforeEach(() => s3.putObjectAsync({
+            Bucket: bucketName,
+            Key: objectName,
+            Body: 'I am the best content ever',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: Nice

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+            return s3.headObjectAsync({ Bucket: bucketName, Key: objectName });
+        }).then(res => {
+            lastModified = res.LastModified;
+        }));
+
+        afterEach(() => bucketUtil.empty(bucketName));
+
+        after(() => bucketUtil.deleteOne(bucketName));
+
+        it('If-Match: returns no error when ETag match, with "', done => {
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: with double quotes around ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+            requestHead({ IfMatch: etag }, err => {
+                checkNoError(err);
+                done();
+            });
+        });
+
+        it('If-Match: returns no error when one of ETags match, with "',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: with double quotes around ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            done => {
+                requestHead({ IfMatch: `non-matching,${etag}` }, err => {
+                    checkNoError(err);
+                    done();
+                });
+            });
+
+        it('If-Match: returns no error when ETag match, without "', done => {
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: without double quotes around ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            requestHead({ IfMatch: etagTrim }, err => {
+                checkNoError(err);
+                done();
+            });
+        });
+
+        it('If-Match: returns no error when one of ETags match, without "',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: without double quotes around ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            });
+
+        it('If-None-Match: returns no error when ETag does not match', done => {
+            requestHead({ IfNoneMatch: 'non-matching' }, err => {
+                checkNoError(err);
+                done();
+            });
+        });
+
+        it('If-None-Match: returns no error when all ETags does not match',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: does s/b do

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            done => {
+                requestHead({
+                    IfNoneMatch: 'non-matching,non-matching-either',
+                }, err => {
+                    checkNoError(err);
+                    done();
+                });
+            });
+
+        it('If-None-Match: returns NotModified when ETag match, with "',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: with double quotes around the ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+            done => {
+                requestHead({ IfNoneMatch: etag }, err => {
+                    checkError(err, 'NotModified');
+                    done();
+                });
+            });
+
+        it('If-None-Match: returns NotModified when one of ETags match, with "',
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: with double quotes around the ETag

File: tests/functional/aws-node-sdk/test/object/objectHead.js [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+
+        it('If-None-Match: returns NotModified when ETag match, without "',
+            done => {
+                requestHead({ IfNoneMatch: etagTrim }, err => {
+                    checkError(err, 'NotModified');
+                    done();
+                });
+            });
+
+        it('If-None-Match: returns NotModified when one of ETags match, ' +
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectHead.js]: without double quotes around ETag

**Comment by LaurenSpiegel** [comment]
Best pr ever 👍

**Comment by alexandre-merle** [comment]
Updated :)

**Comment by rahulreddy** [comment]
👍

**Comment by LaurenSpiegel** [comment]
@ironman-machine thoughts on the matter?

**Comment by alexandre-merle** [comment]
:+1: (for Lauren part ^^)

**Comment by LaurenSpiegel** [comment]
end to end passed

**Pull request merged**